experimental/lib/Support/BooleanLogic/Parser.cpp

```diff
@@ -0,0 +1,215 @@
+//===- Parser.cpp - Parser for boolean expressions --------------*- C++ -*-===//
+//
+// Dynamatic is under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the Parser class, which is responsible for parsing a
+// tokenized boolean logic expression into a boolean expression tree.
+// The parser uses an operator precedence table to manage parsing rules and
+// handles logical operators (AND, OR, NOT) and parentheses. The parsing process
+// involves constructing nodes for the expression tree and reducing sequences of
+// tokens based on precedence rules.
+//
+//===----------------------------------------------------------------------===//
+
+#include "experimental/Support/BooleanLogic/Parser.h"
+#include "experimental/Support/BooleanLogic/BoolExpression.h"
+#include "experimental/Support/BooleanLogic/Lexer.h"
+#include <array>
+#include <cassert>
+
+using namespace dynamatic::experimental::boolean;
+
+enum class Compare : int { lessThan, greaterThan, equal, error, accept };
+
+// Define the static constexpr precedence table
+static constexpr std::array<std::array<Compare, 7>, 7> PRECEDENCE_TABLE = {
+    {{Compare::greaterThan, Compare::greaterThan, Compare::greaterThan,
+      Compare::lessThan, Compare::greaterThan, Compare::lessThan,
+      Compare::greaterThan},
+     {Compare::lessThan, Compare::greaterThan, Compare::greaterThan,
+      Compare::lessThan, Compare::greaterThan, Compare::lessThan,
+      Compare::greaterThan},
+     {Compare::lessThan, Compare::lessThan, Compare::greaterThan,
+      Compare::lessThan, Compare::greaterThan, Compare::lessThan,
+      Compare::greaterThan},
+     {Compare::lessThan, Compare::lessThan, Compare::lessThan,
+      Compare::lessThan, Compare::equal, Compare::lessThan, Compare::error},
+     {Compare::greaterThan, Compare::greaterThan, Compare::greaterThan,
+      Compare::error, Compare::greaterThan, Compare::error,
+      Compare::greaterThan},
+     {Compare::greaterThan, Compare::greaterThan, Compare::greaterThan,
+      Compare::error, Compare::greaterThan, Compare::error,
+      Compare::greaterThan},
+     {Compare::lessThan, Compare::lessThan, Compare::lessThan,
+      Compare::lessThan, Compare::error, Compare::lessThan, Compare::accept}}};
+
+// returns a dynamically-allocated variable
+BoolExpression *dynamatic::experimental::boolean::constructNodeOperator(
+    StackNode *operate, StackNode *s1, StackNode *s2) {
+  assert(operate != nullptr && s1 != nullptr && s2 != nullptr);

```

Comment:
We usually include an error message with asserts, as follows.

```cpp
assert((operate && s1 && s2) && "error message")
```

Same comment applies to asserts below.